zebra-network/src/peer/connection.rs

```diff
@@ -91,34 +84,60 @@ impl Handler {
                 }
             }
             (Handler::Peers, Message::Addr(addrs)) => Handler::Finished(Ok(Response::Peers(addrs))),
+            // `zcashd` returns transactions contiguously (but not necessarily immediately).
+            // It uses `NotFound` if any transactions are missing:
+            // https://github.com/zcash/zcash/blob/e7b425298f6d9a54810cb7183f00be547e4d9415/src/main.cpp#L5617
             (
                 Handler::TransactionsByHash {
                     mut hashes,
                     mut transactions,
                 },
                 Message::Tx(transaction),
             ) => {
+                // assumptions:
+                //   - the transaction list is contiguous
+                //   - missing transaction hashes are included in a `NotFound` message
                 if hashes.remove(&transaction.hash()) {
+                    // we are in the middle of the contiguous transaction response
                     transactions.push(transaction);
+                    if hashes.is_empty() {
+                        Handler::Finished(Ok(Response::Transactions(transactions)))
+                    } else {
+                        Handler::TransactionsByHash {
+                            hashes,
+                            transactions,
+                        }
+                    }
                 } else {
+                    // we are before or after the contiguous block response
                     ignored_msg = Some(Message::Tx(transaction));
-                }
-                if hashes.is_empty() {
-                    Handler::Finished(Ok(Response::Transactions(transactions)))
-                } else {
-                    Handler::TransactionsByHash {
-                        hashes,
-                        transactions,
+                    if !transactions.is_empty() {
+                        warn!("unexpected transaction from peer: transaction responses should be contiguous, followed by notfound. Using partial received transactions as the peer response");
```

Comment:
Change to error to catch bad `zcashd` behaviour:
```suggestion
                        error!("unexpected transaction from peer: transaction responses should be contiguous, followed by notfound. Using partial received transactions as the peer response");
```